### PR TITLE
Feature/setup sql alchemy

### PR DESCRIPTION
**Pull request by kevintrinh388**
finished setting up sqlalchemy

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,6 +10,8 @@ Install requirements (Flask)
 
 `flask run`
 
+You gun need to run database locally cus tech lead don't wanna make a database for me
```
Comment (kphan32) [inline README.md:13]: 🤔

File: app.py [hunk 2]
```diff
@@ -1,12 +1,43 @@
...
   return "hello world"
 
+@app.route("/users")
+def users():
+
+  # test_pass = "testpass"
+  # hashed_pass = hashlib.sha256(test_pass.encode('utf-8')).hexdigest()
+  # test_user = User(email="testemail",password=hashed_pass)
+  # db.session.add(test_user)
+  # db.session.commit()
```
Comment (kphan32) [inline app.py:28]: 🤔

File: app.py [hunk 3]
```diff
@@ -1,12 +1,43 @@
...
 import os
+import hashlib
 
 app = Flask(__name__)
+app.config["SQLALCHEMY_DATABASE_URI"] = os.getenv("DATABASE_URL")
+
+db.init_app(app)
+
+with app.app_context():
+    db.create_all()
```
Comment (kphan32) [inline app.py:13]: need to create db migrations-- i think this will just create tables as needed which isnt what we want

File: app.py [hunk 4]
```diff
@@ -1,12 +1,43 @@
...
+      "password":user.password
+    }
+    res.append(dict)
+
+  return jsonify(res)
+
 if __name__ == "__main__":
   port = int(os.environ.get('PORT', 8080))
-  app.run(debug=False, host='0.0.0.0', port=port)
+  app.run(debug=True, host='0.0.0.0', port=port)
```
Comment (kphan32) [inline app.py:43]: ideally this is only run in prod, so we don't want `Debug` to be true (this block of code isnt run when running `flask run`)

File: app.py [hunk 5]
```diff
@@ -1,12 +1,43 @@
...
+  # test_user = User(email="testemail",password=hashed_pass)
+  # db.session.add(test_user)
+  # db.session.commit()
+
+  users = User.query.all()
+  res = []
+  for user in users:
+    dict = {
+      "email":user.email,
+      "password":user.password
```
Comment (kphan32) [inline app.py:35]: nit: formatting. we might add some linter/formatter later

File: database.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
+"""Handles database connection"""
+import os
+from dotenv import load_dotenv, find_dotenv
+from flask import Flask
+from flask_sqlalchemy import SQLAlchemy
+
+
+load_dotenv(find_dotenv())
```
Comment (kphan32) [inline database.py:8]: this should go into `app`

File: database.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+
+app = Flask(__name__)  # comming from flask library
+
+db_url = os.getenv("DATABASE_URL")
+if db_url.startswith("postgres://"):
+    db_url = db_url.replace("postgres://", "postgresql://", 1)
```
Comment (kphan32) [inline database.py:15]: i think the `replace` call will do nothing if it doesnt start with it, so the if might be redundant

File: models.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+import json
+from database import db
+
+
+class User(db.Model):
+    """User model"""
+
+    id = db.Column(db.Integer, primary_key=True)
+    email = db.Column(db.String(80), unique=True, nullable=False)
+    password = db.Column(db.String(256), nullable=True)
```
Comment (kphan32) [inline models.py:11]: why is the password nullable 🤔

File: models.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+    def __repr__(self):
+        return f"<User {self.email}>"
+
+    def __str__(self):
+        return f"<User {self.email}>"
+
+
+
+db.create_all()
```
Comment (kphan32) [inline models.py:21]: does this get called twice?

File: README.md (hunk 1 above)
Comment (kphan32) [reply README.md:13]: should also leave some instructions on creating the env file. i personally create a `.env.example` file that will be committed with an example of what env vars should be setup. also modify the `.gitignore` to ignore the `.env` file

**Review by kphan32 — CHANGES_REQUESTED** [review]
🤔

File: app.py (hunk 3 above)
Comment (kevintrinh388) [reply app.py:13]: i still dont know what this means btw

Comment (kevintrinh388) [reply app.py:13]: nvm using flask-migrate i tink i got it

File: README.md [hunk 10]
```diff
@@ -2,12 +2,23 @@
 
 ## Setup
 
+Make a .env file in the root folder
+Put this in there
+`export DATABASE_URL="PUT YOUR LOCAL DATABASE URL HERE"`
```
Comment (kphan32) [inline README.md:7]:
```suggestion
`DATABASE_URL="PUT YOUR LOCAL DATABASE URL HERE"`
```

File: app.py [hunk 11]
```diff
@@ -1,12 +1,46 @@
...
+from models import User
+from dotenv import load_dotenv, find_dotenv
+from flask_sqlalchemy import SQLAlchemy
+from flask_migrate import Migrate
 import os
 
 app = Flask(__name__)
+app.config["SQLALCHEMY_DATABASE_URI"] = os.getenv("DATABASE_URL")
+
+load_dotenv(find_dotenv())
```
Comment (kphan32) [inline app.py:11]: need to load dotenv first before using env vars

File: app.py [hunk 12]
```diff
@@ -1,12 +1,46 @@
...
+
+db_url = os.getenv("DATABASE_URL")
+db_url = db_url.replace("postgres://", "postgresql://", 1)
+
+app.config["SQLALCHEMY_DATABASE_URI"] = db_url
+app.config["SQLALCHEMY_TRACK_MODIFICATIONS"] = False
```
Comment (kphan32) [inline app.py:17]: 🤔

File: migrations/versions/a4098101f6f9_initial_migration.py [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+def downgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table('user',
+    sa.Column('id', sa.INTEGER(), autoincrement=True, nullable=False),
+    sa.Column('email', sa.VARCHAR(length=80), autoincrement=False, nullable=False),
+    sa.Column('password', sa.VARCHAR(length=256), autoincrement=False, nullable=True),
+    sa.PrimaryKeyConstraint('id', name='user_pkey'),
+    sa.UniqueConstraint('email', name='user_email_key')
+    )
```
Comment (kphan32) [inline migrations/versions/a4098101f6f9_initial_migration.py:33]: these look backwards

Comment (kphan32) [inline migrations/versions/a4098101f6f9_initial_migration.py:33]: nit: formatting

File: migrations/versions/a4098101f6f9_initial_migration.py [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    op.drop_table('user')
+    # ### end Alembic commands ###
+
+
+def downgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table('user',
+    sa.Column('id', sa.INTEGER(), autoincrement=True, nullable=False),
+    sa.Column('email', sa.VARCHAR(length=80), autoincrement=False, nullable=False),
+    sa.Column('password', sa.VARCHAR(length=256), autoincrement=False, nullable=True),
```
Comment (kphan32) [inline migrations/versions/a4098101f6f9_initial_migration.py:30]: why is password nullable 🤔

**Review by kphan32 — CHANGES_REQUESTED** [review]

**Comment by kphan32** [comment]
[Setup SQL Alchemy](https://trello.com/c/Klz9sVSF/6-setup-sql-alchemy)

**Review by kphan32 — APPROVED** [review]
lgtm :+1:

**Comment by kevintrinh388** [comment]
[Setup SQL Alchemy](https://trello.com/c/Klz9sVSF/6-setup-sql-alchemy)

**Pull request merged**